### PR TITLE
Fix Determinism

### LINKED ISSUE #145: Outputs from runs with same random seed are not identical
**Description of the bug**

I have been unable to get reproducible results when using the same
seed for the random number generators.

**Code example**

Starting from the example described at

  https://stable-baselines.readthedocs.io/en/master/modules/ppo1.html

I can create

```
import gym

from stable_baselines.common.policies import MlpPolicy, MlpLstmPolicy, MlpLnLstmPolicy
from stable_baselines.common.vec_env import DummyVecEnv
from stable_baselines import PPO1

env = gym.make('CartPole-v1')
env = DummyVecEnv([lambda: env])

model = PPO1(MlpPolicy, env, verbose=1)
model.learn(total_timesteps=5000,seed=100)
model.save("ppo1_cartpole")
```
Note that I have added `seed=100` to `model.learn()`.

Running this example prints output to the screen and writes the
ppo1_cartpole.pkl file.

Running the exact same code twice (with the same seed value) produces
different screen outputs and different ppo1_cartpole.pkl files.

**System Info**
My environment:
 * Installed by pip into virtual environment.
 * Stable Baselines version 2.3.0
 * Python version 3.5.2 is installed in the virtual environment.
 * Tensorflow version 1.12.0.
 * OpenAI Gym version 0.10.9.
 * My OS is Ubuntu-16.04.
 * No GPUs.

**Additional context**

It appears from the code that when `seed` is not `None` in `learn()` the
function `set_global_seeds(seed)` is called.  I can see that this
function initialises the following random number generators with the
specified seed:
```
def set_global_seeds(seed):
    """
    set the seed for python random, tensorflow, numpy and gym spaces

    :param seed: (int) the seed
    """
    tf.set_random_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
    gym.spaces.prng.seed(seed)
```

Because of this I also tried including the code lines
```
from stable_baselines.common import set_global_seeds
set_global_seeds(100)
```
before the call to `gym.make()` in the above example, but it did not help.

### LINKED ISSUE #285: Does set_global seed work? [question]
Hello,

I've been playing around with the example code you've provided on the docs and noticed the global seed example in the cart pole environment isn't reproducing the same results on successive runs. 

Is this to be expected?

I've tested this with a fresh install of gym and stable-baselines. 

Thanks

**Code example**
```python
import gym
import numpy as np

from stable_baselines.common.policies import MlpPolicy
from stable_baselines.common.vec_env import SubprocVecEnv
from stable_baselines.common import set_global_seeds
from stable_baselines import ACKTR

def make_env(env_id, rank, seed=0):
    """
    Utility function for multiprocessed env.

    :param env_id: (str) the environment ID
    :param num_env: (int) the number of environments you wish to have in subprocesses
    :param seed: (int) the inital seed for RNG
    :param rank: (int) index of the subprocess
    """
    def _init():
        env = gym.make(env_id)
        env.seed(seed + rank)
        return env
    set_global_seeds(seed)
    return _init

env_id = "CartPole-v1"
num_cpu = 4  # Number of processes to use
# Create the vectorized environment
env = SubprocVecEnv([make_env(env_id, i) for i in range(num_cpu)])

model = ACKTR(MlpPolicy, env, verbose=1)
model.learn(total_timesteps=25000)

obs = env.reset()
for _ in range(1000):
    action, _states = model.predict(obs)
    obs, rewards, dones, info = env.step(action)
    env.render()

```

### PR DESCRIPTION
**Pull request by araffin**
closes #145
closes #285
Related #485

**Important Note**: In order to fix determinism, I had to break the API of `learn()` method: I moved the `seed` argument to the constructor.
I also had to separate the determinism test to avoid side effects from other tests

### CODE REVIEW

**Comment by araffin** [comment]
>in which case we do not set the seed to anything fixed. 
Done.

>self.seed should be stored in the save file. Looks like just adding it to the list of stored variables should do the trick.

Good point ;) (I had that thought afterward too)

>TD3 test is failing for me with current version
>Sidenote: Other than the TD3, other algorithms pass the test

I think we need to investigate more, TD3 has been the one giving me the more problems. Weirdly enough, it seems to be affected by previous execution of models... (running the test with TD3 alone was different from running TD3 after DQN test)
The thing I don't really understand is that TD3 code is based on SAC but SAC does not fail... I would appreciate some help ;)

The tests pass on travis... What is your tf version? Did you try with the docker image?

>should skip tests if GPU is used 

Do you have a code snippet for checking that automatically?
And do you have some pointers to solve that issue properly? (I know it is possible with pytorch but I think for tf, it won't be possible before tf 2.0)

>Nitpick: What does 0deterministic stand for in

yes, that's a hack for having a separate test, otherwise, it would fail...I'm open to any better solution as I don't like that one.

**Comment by Miffyli** [comment]
>I think we need to investigate more, TD3 has been the one giving me the more problems. Weirdly enough, it seems to be affected by previous execution of models... (running the test with TD3 alone was different from running TD3 after DQN test)

Now that you mentioned one of the runs randomly failed with both DQN and TD3 failing out of the blue :o . I can try digging in further to this, but I am currently connecting over mobile connections and SSHing is no fun, so things will be tad slow ^^.

> The tests pass on travis... What is your tf version? Did you try with the docker image?

Tensorflow 1.14.0 without docker image (I test things without docker as that is where I would personally run things). This is probably the part that causes issues as stable-baselines is based on the older TF version.

> Do you have a code snippet for checking that automatically?

I do not know about detection but you could possibly run test with [`with tf.device("cpu")`](https://www.tensorflow.org/guide/gpu#manual_device_placement) block to force running things on CPU. As for fixing the issue, I think we should move that to TF2/next-backend thing, given that TF2 now is officially released.

> yes, that's a hack for having a separate test, otherwise, it would fail...I'm open to any better solution as I don't like that one.

Aight :). I do not know enough of these automatic systems to be of any help.

File: stable_baselines/a2c/a2c.py [hunk 1]
```diff
@@ -38,14 +38,21 @@ class A2C(ActorCriticRLModel):
     :param policy_kwargs: (dict) additional arguments to be passed to the policy on creation
     :param full_tensorboard_log: (bool) enable additional logging when using tensorboard
         WARNING: this logging can take a lot of space quickly
+    :param seed: (int) Seed for the pseudo-random generators (python, numpy, tensorflow).
+        Note that if you want completely deterministic results, you must set
```
Comment (Miffyli) [inline stable_baselines/a2c/a2c.py]: Docs on `seed = None` (random seed and `None` is default)

**Comment by araffin** [comment]
I did additional tests and it seems it comes from tensorflow, I could not reset completely its state and reload is not supported... so I'm a bit stuck (see [issue](https://github.com/tensorflow/tensorflow/issues/15741))

**Comment by Miffyli** [comment]
Sounds like something that also can be better fixed along with TF2 support? At this rate we need a detailed list of things to consider when updating to the new backend :) (I know few other suggestions apart from determinism etc).

**Comment by araffin** [comment]
>Sounds like something that also can be better fixed along with TF2 support? 

I hope so. Referencing #366  then.

But I think it still makes sense to merge that partial fix, no? or you prefer to wait for a full patch?

**Comment by Miffyli** [comment]
Oh yes, definitely should merge this one, too :). It is a clear step to right direction. I would only add a notification that TD3 might not work as expected but for other algorithms things work as intended.

**Comment by bycn** [comment]
Any updates for the TD3 failures? I finally found this thread and am experiencing the exact same issues — it seems to randomly fail AFTER a certain sequence of runs. 

Also, is DDPG affected by this?
